src/box/memtx_engine.cc

```diff
@@ -202,6 +214,80 @@ memtx_build_secondary_keys(struct space *space, void *param)
 	return 0;
 }
 
+/**
+ * Secondary indexes can be built in bulk with O(n) sorting algorithm after the
+ * PK data is recovered (after the initial recovery), if we have the memtx sort
+ * data enabled and available. This function builds such indexes.
+ */
+static int
+memtx_build_presorted_secondary_keys(struct space *space, void *param)
+{
+	uint32_t space_id = space->def->id;
+	struct memtx_space *memtx_space = (struct memtx_space *)space;
+	if (space->engine != param || space_index(space, 0) == NULL ||
+	    memtx_space->replace == memtx_space_replace_all_keys ||
+	    space_id_is_system(space_id))
+		return 0;
+
+	/* We must have memtx index sort data available. */
+	struct memtx_engine *memtx = (struct memtx_engine *)space->engine;
+	assert(memtx->msdr != NULL);
+
+	/*
+	 * A before_replace trigger can change the inserted tuple, so we should
+	 * not proceed using the O(n) SK sort that does not account that.
+	 */
+	if (space_has_before_replace_triggers(space)) {
```

Comment:
It seems to me the fact we have a `before_replace` trigger on a space does not break the new sorting approach. The tuple is updated on the MemTX side and we receive it in the TREE index in its final form (after `before_replace` trigger has been fired), so we can save the new tuple in the old to new tuple address map and successfully translate the old pointer from the snapshot to the tuple resulted in `before_replace`.

Reply:
Removed.

Reply:
Dropped (to be pushed).